### PR TITLE
Support undecorated functions

### LINKED ISSUE #202: Push off gtscript.function external symbol resolution time until stencil generation time
In fv3core, for various reasons, we manually invoke `gtscript.stencil(definition=stencil_func, ...)` at the time of first usage, since we are still setting up the backend and rebuild flag at run-time in the tests. However, using externals that are defined later than module-load-time are not working, since `gtscript.function` tries to resolve the externals at load time.

I propose changing the decorator to register functions, and delay the symbol resolution until the time the stencil inlines the call.

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Support using undecorated functions. Delays external resolution for these.

Resolves #202.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Review by egparedes — COMMENTED** [review]
Since it is a very minimal PR I don't have a lot to add. My only point is that the testing is a bit too concise. Could you improve the tests to also check that non-explicit external variables use the right value if they are modified between the definition and the compilation points ? Something similar in spirit to the example in issue 202: https://github.com/GridTools/gt4py/issues/202#issuecomment-709415546

**Comment by jdahm** [comment]
In making the more comprehensive test, I found out that if the function is imported as an external, then the resolution is not done correctly yet in this PR. This will take a bit more effort to get the annotation done in the correct spot in the code.

File: src/gt4py/frontend/gtscript_frontend.py [hunk 1]
```diff
@@ -1230,7 +1230,7 @@ def collect_external_symbols(definition):
         imported_symbols = {name: {} for name in imported_names}
 
         context, unbound = gt_meta.get_closure(
-            definition, included_nonlocals=False, include_builtins=False
+            definition, included_nonlocals=True, include_builtins=False
```
Comment (jdahm) [inline src/gt4py/frontend/gtscript_frontend.py:1233]: `True` is needed here to be able to include the value of `A` in `inspect.closurevars` when importing the function (which is using an external `A`) as an external.

File: src/gt4py/frontend/gtscript_frontend.py [hunk 2]
```diff
@@ -1262,11 +1262,17 @@ def collect_external_symbols(definition):
         return nonlocal_symbols, imported_symbols
 
     @staticmethod
-    def eval_constant(name: str, context: dict, loc=None):
+    def eval_external(name: str, context: dict, loc=None):
```
Comment (jdahm) [inline src/gt4py/frontend/gtscript_frontend.py:1265]: Renamed `eval_constant` to `eval_external` to better indicate the scope of what it performs for the frontend.

File: src/gt4py/frontend/gtscript_frontend.py (hunk 1 above)
Comment (jdahm) [reply src/gt4py/frontend/gtscript_frontend.py:1401]: cc: @tehrengruber

**Comment by jdahm** [comment]
> My only point is that the testing is a bit too concise. Could you improve the tests to also check that non-explicit external variables use the right value if they are modified between the definition and the compilation points ?

Agreed. I added this. It is now ready to review again. Thanks!

File: src/gt4py/frontend/gtscript_frontend.py (hunk 1 above)
Comment (egparedes) [reply src/gt4py/frontend/gtscript_frontend.py:1401]: I can't tell anymore but it feels to me that this change could affect the current behavior of the frontend in other places. Have you considered and tested all the implications? @tehrengruber do you also agree this is a safe change?

File: tests/test_unittest/test_gtscript_frontend.py [hunk 3]
```diff
@@ -135,6 +135,46 @@ def definition_func(inout_field: gtscript.Field[float]):
...
+            with computation(PARALLEL), interval(...):
+                inout_field = func()
+
+        stencil_id, def_ir = compile_definition(
+            definition_func, "test_undecorated_delay", module, externals=externals
+        )
+
+        stmt = def_ir.computations[0].body.stmts[0]
+        assert isinstance(stmt.value, gt_ir.ScalarLiteral) and stmt.value.value == 1
+
```
Comment (egparedes) [inline tests/test_unittest/test_gtscript_frontend.py:177]: I think a test is missing here to check that the decorated function will freeze the value of A even if it is changed later.

**Review by egparedes — COMMENTED** [review]
Thanks for adding the tests. It looks almost ready now, just a question and a minor test extension.

File: tests/test_unittest/test_gtscript_frontend.py [hunk 4]
```diff
@@ -148,7 +188,7 @@ def definition_func(inout_field: gtscript.Field[float]):
             with computation(PARALLEL), interval(...):
                 inout_field = inout_field[0, 0, 0] + WRONG_VALUE_CONSTANT
 
-        with pytest.raises(gt_frontend.GTScriptSymbolError, match=r".*WRONG_VALUE_CONSTANT.*"):
+        with pytest.raises(gt_frontend.GTScriptDefinitionError, match=r".*WRONG_VALUE_CONSTANT.*"):
             compile_definition(definition_func, "test_wrong_value", module, externals=externals)
```
Comment (jdahm) [inline tests/test_unittest/test_gtscript_frontend.py:192]: @egparedes This was the only place I could tell that the behavior changed.

Comment (egparedes) [reply tests/test_unittest/test_gtscript_frontend.py:192]: Ok, then, raising a different Exception should not matter too much.

File: tests/test_unittest/test_gtscript_frontend.py (hunk 3 above)
Comment (egparedes) [reply tests/test_unittest/test_gtscript_frontend.py:177]: Once this new test is added, I will approve the PR

Comment (jdahm) [reply tests/test_unittest/test_gtscript_frontend.py:177]: I'll fix that tonight. Thanks for reminding me.

**Comment by jdahm** [comment]
@egparedes Apologies for the delay. This is ready to go!

**Review by egparedes — APPROVED** [review]
Thanks for the changes. This is now ready to go.

**Pull request merged**